File: src/main/kotlin/io/github/dylmeadows/eontimer/service/factory/timer/EntralinkTimer.kt
package io.github.dylmeadows.eontimer.service.factory.timer

import io.github.dylmeadows.eontimer.util.milliseconds
import io.github.dylmeadows.eontimer.util.reactor.TimerState
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import java.time.Duration

@Service
class EntralinkTimer @Autowired constructor(
    private val delayTimer: DelayTimer) {

    fun createStages(targetSecond: Long, targetDelay: Long, entralinkCalibration: Long, calibration: Long): List<Duration> {
        val stages = delayTimer.createStages(targetSecond, targetDelay, calibration)
        return listOf(
            stage1(stages),
            stage2(stages, entralinkCalibration))
    }

    fun createTimer(targetSecond: Long, targetDelay: Long, entralinkCalibration: Long, calibration: Long): Flux<TimerState> {
        // TODO: fix this
        return Flux.empty()
    }

    fun calibrate(targetDelay: Long, delayHit: Long): Long {
        return delayTimer.calibrate(targetDelay, delayHit)
    }

    private fun stage1(stages: List<Duration>): Duration {
        return stages[0] + 250L.milliseconds
    }

    private fun stage2(stages: List<Duration>, entralinkCalibration: Long): Duration {
        return stages[1] - entralinkCalibration.milliseconds
    }
}

File: src/services/timers/EntralinkTimer.cpp
//
// Created by Dylan Meadows on 2020-03-20.
//

#include "EntralinkTimer.h"

namespace service::timer {
    EntralinkTimer::EntralinkTimer(const DelayTimer *delayTimer)
        : delayTimer(delayTimer) {
    }

    const std::shared_ptr<std::vector<int>>
    EntralinkTimer::createStages(const int targetDelay, const int targetSecond, const int calibration, const int entralinkCalibration) const {
        std::shared_ptr<std::vector<int>> stages = std::make_shared<std::vector<int>>(2);
        (*stages)[0] = createStage1(targetDelay, targetSecond, calibration);
        (*stages)[1] = createStage2(targetDelay, calibration, entralinkCalibration);
        return stages;
    }

    int EntralinkTimer::createStage1(const int targetDelay, const int targetSecond, const int calibration) const {
        return delayTimer->createStage1(targetDelay, targetSecond, calibration) + 250;
    }

    int EntralinkTimer::createStage2(const int targetDelay, const int calibration, const int entralinkCalibration) const {
        return delayTimer->createStage2(targetDelay, calibration) - entralinkCalibration;
    }

    int EntralinkTimer::calibrate(const int targetDelay, const int delayHit) const {
        return delayTimer->calibrate(targetDelay, delayHit);
    }
}
